src/test/java/com/google/api/codegen/testdata/csharp_gapic_resourcenames_library.baseline

```diff
@@ -0,0 +1,685 @@
+        /// <param name="bookId">The book ID. Must not be <c>null</c>.</param>
+        /// <param name="returnId">The return ID. Must not be <c>null</c>.</param>
+        public ReturnName(string shelfId, string bookId, string returnId)
+        {
+            ShelfId = GaxPreconditions.CheckNotNull(shelfId, nameof(shelfId));
+            BookId = GaxPreconditions.CheckNotNull(bookId, nameof(bookId));
+            ReturnId = GaxPreconditions.CheckNotNull(returnId, nameof(returnId));
+        }
+
+        /// <summary>
+        /// The shelf ID. Never <c>null</c>.
+        /// </summary>
+        public string ShelfId { get; }
+
+        /// <summary>
+        /// The book ID. Never <c>null</c>.
+        /// </summary>
+        public string BookId { get; }
+
+        /// <summary>
+        /// The return ID. Never <c>null</c>.
+        /// </summary>
+        public string ReturnId { get; }
+
+        /// <inheritdoc />
+        public override string ToString() => s_template.Expand(ShelfId, BookId, ReturnId);
+    }
+
+    public sealed partial class BookNameOneof : IResourceName
+    {
+        public enum Type
+        {
+            BOOK_NAME = 1,
+            ARCHIVED_BOOK_NAME = 2,
+            DELETED_BOOK_NAME_FIXED = 3,
+        }
+
+        public static BookNameOneof Parse(string name)
+        {
+            BookNameOneof result;
+            if (TryParse(name, out result))
+            {
+                return result;
+            }
+            throw new ArgumentException("Invalid name", "name");
+        }
+
+        public static bool TryParse(string name, out BookNameOneof result)
+        {
+            BookName bookName;
+            if (BookName.TryParse(name, out bookName))
+            {
+                result = new BookNameOneof(Type.BOOK_NAME, bookName);
+                return true;
+            }
+            ArchivedBookName archivedBookName;
+            if (ArchivedBookName.TryParse(name, out archivedBookName))
+            {
+                result = new BookNameOneof(Type.ARCHIVED_BOOK_NAME, archivedBookName);
+                return true;
+            }
+            DeletedBookNameFixed deletedBookNameFixed;
+            if (DeletedBookNameFixed.TryParse(name, out deletedBookNameFixed))
+            {
+                result = new BookNameOneof(Type.DELETED_BOOK_NAME_FIXED, deletedBookNameFixed);
+                return true;
+            }
+            result = null;
+            return false;
+        }
+
+        public static BookNameOneof From(BookName bookName) => new BookNameOneof(Type.BOOK_NAME, bookName);
+
+        public static BookNameOneof From(ArchivedBookName archivedBookName) => new BookNameOneof(Type.ARCHIVED_BOOK_NAME, archivedBookName);
+
+        public static BookNameOneof From(DeletedBookNameFixed deletedBookNameFixed) => new BookNameOneof(Type.DELETED_BOOK_NAME_FIXED, deletedBookNameFixed);
+
+        public BookNameOneof(Type type, IResourceName value)
```

Comment:
Validate that `type` is a member of the enum?
Ideally, validate that the type of the value corresponds to the type, as well... we could do that via reflection, having an attribute on each enum value. Definitely a later enhancement.